src/main/java/org/sonarlint/intellij/config/global/ServerConnectionService.kt

```diff
@@ -0,0 +1,159 @@
+    }
+
+    fun getServerConnectionByName(name: String): Optional<ServerConnection> {
+        return Optional.ofNullable(getConnections().firstOrNull { name == it.name })
+    }
+
+    fun connectionExists(connectionName: String): Boolean {
+        return getConnections().any { it.name == connectionName }
+    }
+
+    fun getServerNames(): Set<String> {
+        return getConnections().map { it.name }.toSet()
+    }
+
+    fun addServerConnection(connection: ServerConnection) {
+        setServerConnections(getConnections() + connection)
+    }
+
+    fun replaceConnection(name: String, replacementConnection: ServerConnection) {
+        val serverConnections = getConnections().toMutableList()
+        serverConnections[serverConnections.indexOfFirst { it.name == name }] = replacementConnection
+        setServerConnections(serverConnections.toImmutableList())
+    }
+
+    private fun setServerConnections(connections: List<ServerConnection>) {
+        setServerConnections(getGlobalSettings(), connections)
+    }
+
+    fun setServerConnections(settings: SonarLintGlobalSettings, connections: List<ServerConnection>) {
+        val previousConnections = getConnections()
+        settings.serverConnections = connections.map {
+            var builder = ServerConnectionSettings.newBuilder().setName(it.name).setHostUrl(it.hostUrl).setDisableNotifications(it.notificationsDisabled)
+            if (it is SonarCloudConnection) {
+                builder = builder.setOrganizationKey(it.organizationKey)
+            }
+            builder.build()
+        }.toList()
+        connections.forEach { saveCredentials(it.name, it.credentials) }
+        notifyConnectionsChange(connections)
+        notifyCredentialsChange(previousConnections, connections)
+        val removedConnectionNames = previousConnections.map { it.name }.filter { name -> !connections.map { it.name }.contains(name) }
+        removedConnectionNames.forEach { forgetCredentials(it) }
+    }
+
+    private fun notifyConnectionsChange(connections: List<ServerConnection>) {
+        ApplicationManager.getApplication().messageBus.syncPublisher(ServerConnectionsListener.TOPIC).afterChange(connections)
+    }
+
+    private fun notifyCredentialsChange(previousConnections: List<ServerConnection>, newConnections: List<ServerConnection>) {
+        val changedConnections = newConnections.filter { connection ->
+            val previousConnection = previousConnections.find { it.name == connection.name }
+            previousConnection?.let { connection.credentials != it.credentials } == true
+        }
+        if (changedConnections.isNotEmpty()) {
+            ApplicationManager.getApplication().messageBus.syncPublisher(ServerConnectionsListener.TOPIC).credentialsChanged(changedConnections)
+        }
+    }
+
+    private fun loadCredentials(connectionId: String): ServerConnectionCredentials? {
+        val token = PasswordSafe.instance.getPassword(tokenCredentials(connectionId))
```

Comment:
This is a slow operation and it should not be run on the EDT 😢 

This is especially annoying because it is called from each `IssueNode` when rendering

2 stack traces I got:

```
2023-11-02 11:06:53,403 [  22494] SEVERE - #c.i.u.SlowOperations - Slow operations are prohibited on EDT. See SlowOperations.assertSlowOperationsAreAllowed javadoc.
java.lang.Throwable: Slow operations are prohibited on EDT. See SlowOperations.assertSlowOperationsAreAllowed javadoc.
	at com.intellij.openapi.diagnostic.Logger.error(Logger.java:376)
	at com.intellij.util.SlowOperations.assertNonCancelableSlowOperationsAreAllowed(SlowOperations.java:115)
	at com.intellij.ide.passwordSafe.impl.BasePasswordSafe.get(PasswordSafeImpl.kt:93)
	at com.intellij.credentialStore.CredentialStore.getPassword(CredentialStore.java:30)
	at org.sonarlint.intellij.config.global.ServerConnectionService.loadCredentials(ServerConnectionService.kt:128)
	at org.sonarlint.intellij.config.global.ServerConnectionService.getConnections(ServerConnectionService.kt:52)
	at org.sonarlint.intellij.core.ProjectBindingManager.tryGetServerConnection(ProjectBindingManager.java:148)
	at org.sonarlint.intellij.ui.nodes.IssueNode.retrieveServerConnection(IssueNode.java:59)
	at org.sonarlint.intellij.ui.nodes.IssueNode.doRender(IssueNode.java:71)
	at org.sonarlint.intellij.ui.nodes.IssueNode.lambda$render$0(IssueNode.java:67)
[...]
	at org.sonarlint.intellij.common.ui.ReadActionUtils$Companion.runReadActionSafely(ReadActionUtils.kt:34)
	at org.sonarlint.intellij.common.ui.ReadActionUtils.runReadActionSafely(ReadActionUtils.kt)
	at org.sonarlint.intellij.ui.nodes.IssueNode.render(IssueNode.java:67)
	at org.sonarlint.intellij.ui.tree.TreeCellRenderer.customizeCellRenderer(TreeCellRenderer.java:37)
[...]
	at org.sonarlint.intellij.core.ProjectBindingManager.bindTo(ProjectBindingManager.java:188)
	at org.sonarlint.intellij.config.project.SonarLintProjectSettingsPanel.save(SonarLintProjectSettingsPanel.java:105)
	at org.sonarlint.intellij.config.project.SonarLintProjectConfigurable.apply(SonarLintProjectConfigurable.java:96)
```

```
2023-11-02 11:06:44,520 [  13611] SEVERE - #c.i.u.SlowOperations - Slow operations are prohibited on EDT. See SlowOperations.assertSlowOperationsAreAllowed javadoc.
java.lang.Throwable: Slow operations are prohibited on EDT. See SlowOperations.assertSlowOperationsAreAllowed javadoc.
	at com.intellij.openapi.diagnostic.Logger.error(Logger.java:376)
	at com.intellij.util.SlowOperations.assertNonCancelableSlowOperationsAreAllowed(SlowOperations.java:115)
	at com.intellij.ide.passwordSafe.impl.BasePasswordSafe.get(PasswordSafeImpl.kt:93)
	at com.intellij.credentialStore.CredentialStore.getPassword(CredentialStore.java:30)
	at org.sonarlint.intellij.config.global.ServerConnectionService.loadCredentials(ServerConnectionService.kt:128)
	at org.sonarlint.intellij.config.global.ServerConnectionService.getConnections(ServerConnectionService.kt:52)
	at org.sonarlint.intellij.config.global.ServerConnectionService.connectionExists(ServerConnectionService.kt:76)
	at org.sonarlint.intellij.config.project.SonarLintProjectBindPanel.onConnectionSelected(SonarLintProjectBindPanel.java:158)
	at org.sonarlint.intellij.config.project.SonarLintProjectBindPanel$ServerItemListener.itemStateChanged(SonarLintProjectBindPanel.java:415)
```

Reply:
:rage: I did not face it on my side... Thanks for testing, I will have to rework the PR then, I will ping you when it's done